### DISCUSSION TITLE
Anti-aliased Arc

### DISCUSSION THREAD

**Discussion by hansaplasst** [Bodmer/TFT_eSPI#2248] [New processor/display support or capability request] [answered]
It would be nice to have a function which can draw an anti-aliased arc.
I have tried to implement it. But I lack the skills.

```
// Draw an anti-aliased arc from lowangle to hiangle at x, y with width w and radius r
// If bg_color is not included the background pixel color will be read from TFT or sprite
tft.drawSmoothArc(int16_t lowangle, int16_t hiangle, int32_t x, int32_t y, int32_t w, int32_t r, uint32_t color, uint32_t bg_color = 0x00FFFFFF);
```

**Comment by Bodmer** [comment]
That is one of the more difficult algorithms to implement. The LVGL library implements an arc function but I have not looked at how it is done.  I suspect the filled SmoothCircle function could be adapted but I would have to think about how the endpoints are implemented. The ILI9341_Due library had an arc function but it was not antialiased but could probably be converted.

It is not something I have time to experiemnt with at the moment.

**Reply by hansaplasst** [reply]
Hi Bodmer. Thanks for the info. I will check it out and if I find some suitable code I'll let you know.

**Comment by Bodmer** [comment]
I thought about this further and had a go at implementing it. Here is a demo of the new function:

[deleted file]

The arc is antialiased except at the ends, I will think on that more.

The background is not read at the moment but when added to the library that will be an option.

**Reply by Bodmer** [reply]
Just updated the zip, last minute tweaks caused a compile problem.

**Reply by hansaplasst** [reply]
Wow.. Thanks! Didn't expect a solution so soon!
I compiled it and it looks very promising! I will give it a go in my project and let you know the results (probably after the holidays).

Happy Holidays!
-Hansa

**Reply by hansaplasst** [reply]
> The arc is antialiased except at the ends, I will think on that more.

If you anti-alias the ends, what happens if a part of an existing arc is redrawn? Will there be a deviation in the arc? In that case I would make it optional. 

By the way, on my tiny GC9A01 display, it's barely noticeable that the ends are not anti-aliased.

**Comment by Bodmer** [comment]
The sketch uploaded had rendering errors, so it has been deleted. This has been fixed in my copy and I will add the new function to the library.

**Reply by hansaplasst** [reply]
Great!
When do you commit these changes?

**Comment by hansaplasst** [comment]
Here is my project. It's a work in progress so it needs some cleaning up :broom: :broom: 

[TempController.mp4](https://user-images.githubusercontent.com/27355296/210116486-af1fae57-461e-446c-999a-2be9a092f64a.mp4)

I've attached a pot-meter to TMP_RD (GPIO15) and a wire to TOUCH_RD (GPIO4) for testing purposes.
I redraw the scale including scale-labels. If you make `needle.height = 50` in the setup() this will be more clear.

BTW. The code also works for an ESP8266 if you remove the sprite and ADC stuff.
I will leave a final a version for both MCU's when I'm finished coding.

Thanks for all your help so far!

**Reply by Bodmer** [reply]
Thanks. I have deleted the zip file because it contained an early version of the arc code which has faults and can cause a processor to crash. The library branch mentioned below now contains the updated functions.

**Comment by Bodmer** [comment, marked as answer]
I have added new arc rendering features to the master branch of the library here:

https://github.com/Bodmer/TFT_eSPI

New examples have been added.
